File: packages/btree/umbra.cc
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <mutex>
#include <thread>

class ColumnStore
{
public:
    void insert(const std::string &columnName, const std::string &value)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        auto &column = getColumn(columnName);
        auto newVersion = createNewVersion();

        for (auto &entry : column)
        {
            if (entry.first <= currentVersion_)
            {
                entry.second.push_back(value);
            }
            else
            {
                entry.second.push_back(value); // Clone value if needed
            }
            incrementVersion(entry.first);
        }

        column[newVersion] = {value};
    }

    void bulkInsert(const std::string &columnName, const std::vector<std::string> &values)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        auto &column = getColumn(columnName);
        auto newVersion = createNewVersion();

        for (auto &entry : column)
        {
            if (entry.first <= currentVersion_)
            {
                entry.second.insert(entry.second.end(), values.begin(), values.end());
            }
            else
            {
                entry.second.insert(entry.second.end(), values.begin(), values.end()); // Clone values if needed
            }
            incrementVersion(entry.first);
        }

        column[newVersion] = values;
    }

    void erase(const std::string &columnName, int version)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        auto &column = getColumn(columnName);
        auto ver = getVersion(version);
        auto it = column.find(ver);

        if (it != column.end())
        {
            it->second.pop_back();
            incrementVersion(ver);
        }
    }

    void bulkErase(const std::string &columnName, int version, int count)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        auto &column = getColumn(columnName);
        auto ver = getVersion(version);
        auto it = column.find(ver);

        if (it != column.end())
        {
            auto &entries = it->second;
            if (count >= entries.size())
            {
                entries.clear();
            }
            else
            {
                entries.resize(entries.size() - count);
            }
            incrementVersion(ver);
        }
    }

    std::vector<std::string> get(const std::string &columnName, int version)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        auto &column = getColumn(columnName);
        auto ver = getVersion(version);
        auto it = column.find(ver);
        if (it != column.end())
        {
            return it->second;
        }
        return {};
    }

private:
    std::unordered_map<std::string, std::unordered_map<int, std::vector<std::string>>> columns_;
    std::unordered_map<int, int> versions_;
    int currentVersion_ = 0;
    std::mutex mutex_;

    std::unordered_map<int, std::vector<std::string>> &getColumn(const std::string &columnName)
    {
        return columns_[columnName];
    }

    int getVersion(int version)
    {
        auto it = versions_.find(version);
        if (it == versions_.end())
        {
            it = versions_.insert({version, 0}).first;
        }
        return it->second;
    }

    int createNewVersion()
    {
        int newVersion = currentVersion_ + 1;
        currentVersion_ = newVersion;
        versions_[newVersion] = 0;
        return newVersion;
    }

    void incrementVersion(int version)
    {
        auto it = versions_.find(version);
        if (it != versions_.end())
        {
            it->second++;
        }
    }
};

int main()
{
    ColumnStore store;

    std::thread t1([&store]()
                   {
        store.insert("Column1", "Value1");
        store.insert("Column1", "Value2");
        store.insert("Column1", "Value3");
        store.bulkInsert("Column2", { "ValueA", "ValueB", "ValueC" });
        store.erase("Column1", 2);
        store.bulkErase("Column2", 1, 2); });

    std::thread t2([&store]()
                   {
        store.insert("Column1", "Value4");
        store.insert("Column1", "Value5");
        store.bulkInsert("Column2", { "ValueD", "ValueE", "ValueF" });
        store.erase("Column1", 1);
        store.bulkErase("Column2", 1, 1); });

    t1.join();
    t2.join();

    auto values1 = store.get("Column1", 2);
    std::cout << "Column1 (Version 2): ";
    for (const auto &value : values1)
    {
        std::cout << value << " ";
    }
    std::cout << std::endl;

    auto values2 = store.get("Column2", 2);
    std::cout << "Column2 (Version 2): ";
    for (const auto &value : values2)
    {
        std::cout << value << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: packages/btree/umbra.ts
class ColumnStore {
    columns: Map<string, Map<number, any[]>>;
    versions: Map<number, number>;
    currentVersion: number;

    constructor() {
        this.columns = new Map<string, Map<number, any[]>>();
        this.versions = new Map<number, number>();
        this.currentVersion = 0;
        this.versions.set(this.currentVersion, 0);
    }

    private getColumn(columnName: string): Map<number, any[]> {
        let column = this.columns.get(columnName);
        if (!column) {
            column = new Map<number, any[]>();
            this.columns.set(columnName, column);
        }
        return column;
    }

    private getVersion(version: number): number {
        let ver = this.versions.get(version);
        if (!ver) {
            ver = 0;
            this.versions.set(version, ver);
        }
        return ver;
    }

    private createNewVersion(): number {
        const newVersion = this.currentVersion + 1;
        this.currentVersion = newVersion;
        this.versions.set(newVersion, 0);
        return newVersion;
    }

    private incrementVersion(version: number): void {
        const ver = this.versions.get(version);
        if (ver !== undefined) {
            this.versions.set(version, ver + 1);
        }
    }

    insert(columnName: string, value: any): void {
        const column = this.getColumn(columnName);
        const newVersion = this.createNewVersion();

        for (const [version, entries] of column) {
            if (version <= this.currentVersion) {
                entries.push(value);
            } else {
                const clonedValue = JSON.parse(JSON.stringify(value));
                entries.push(clonedValue);
            }
            this.incrementVersion(version);
        }

        column.set(newVersion, [value]);
    }

    bulkInsert(columnName: string, values: any[]): void {
        const column = this.getColumn(columnName);
        const newVersion = this.createNewVersion();

        for (const [version, entries] of column) {
            if (version <= this.currentVersion) {
                entries.push(...values);
            } else {
                const clonedValues = values.map((value) => JSON.parse(JSON.stringify(value)));
                entries.push(...clonedValues);
            }
            this.incrementVersion(version);
        }

        column.set(newVersion, values.slice());
    }

    delete(columnName: string, version: number): void {
        const column = this.getColumn(columnName);
        const ver = this.getVersion(version);
        const entries = column.get(ver);

        if (entries) {
            entries.pop();
            this.incrementVersion(ver);
        }
    }

    bulkDelete(columnName: string, version: number, count: number): void {
        const column = this.getColumn(columnName);
        const ver = this.getVersion(version);
        const entries = column.get(ver);

        if (entries) {
            entries.splice(-count);
            this.incrementVersion(ver);
        }
    }

    get(columnName: string, version: number): any[] {
        const column = this.getColumn(columnName);
        const ver = this.getVersion(version);
        return column.get(ver) || [];
    }
}
